### PR TITLE
Partial staking onchain logic

### PR DESCRIPTION
**Pull request by kaichaosun**

### CODE REVIEW

File: pallets/staking/src/lib.rs [hunk 1]
```diff
@@ -12,86 +12,149 @@
...
             Balance = Balance,
             Amount = Amount,
         >;
 
-        /// The loan's module id, keep all collaterals of CDPs.
+        /// Currency used for staking
+        #[pallet::constant]
+        type StakingCurrency: Get<CurrencyId>;
+
+        /// Currency used for liquid voucher
```
Comment (GopherJ) [inline pallets/staking/src/lib.rs:68]: I think voucher isn't the same thing with staking mecanism as voucher means `money replacement`, people can spend it.

However, staking just ensure that people can get money back, so I think it's not the same thing

File: pallets/staking/src/lib.rs [hunk 2]
```diff
@@ -111,53 +174,157 @@ pub mod module {
...
+        /// May only be called from `T::WithdrawOrigin`.
+        ///
+        /// - `agent`: the multisig account of relay chain.
+        /// - `amount`: the requested assets.
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn unstake(origin: OriginFor<T>, amount: Balance) -> DispatchResultWithPostInfo {
-            let who = ensure_signed(origin)?;
-            Self::unstake_internal(&who, amount)?;
+        pub fn withdraw(
```
Comment (GopherJ) [inline pallets/staking/src/lib.rs:230]: it looks like that people can stake 1 dot then withdraw 100dot as there is no limit?

File: pallets/staking/src/lib.rs [hunk 3]
```diff
@@ -111,53 +174,157 @@ pub mod module {
...
+            agent: T::AccountId,
+            amount: Balance,
+        ) -> DispatchResultWithPostInfo {
+            T::WithdrawOrigin::ensure_origin(origin)?;
+            ensure!(
+                amount <= T::MaxWithdrawAmount::get(),
+                Error::<T>::ExcessWithdraw
+            );
+
+            T::Currency::transfer(
```
Comment (GopherJ) [inline pallets/staking/src/lib.rs:241]: will this influence exchange rate?  As total staking has changed

File: pallets/staking/src/lib.rs (hunk 2 above)
Comment (kaichaosun) [reply pallets/staking/src/lib.rs:230]: This function can only be called by a specific collective, the collective may be multisig accounts in parachain, which also have a corresponding account in relay chain.

Comment (kaichaosun) [inline pallets/staking/src/lib.rs:230]: A better name.

File: pallets/staking/src/lib.rs [hunk 4]
```diff
@@ -111,53 +174,157 @@ pub mod module {
...
+        /// Record the staking rewards, no real transfer.
+        /// TODO restrict the times an account can report in one day and max rewards.
+        ///
+        /// May only be called from `T::WithdrawOrigin`.
+        ///
+        /// - `amount`: the rewarded assets.
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn return_pending_balance(
+        pub fn record_rewards(
```
[the remark this answers is not in the document]
Comment (kaichaosun) [reply pallets/staking/src/lib.rs:260]: sudo remove rewards operation.

**Review by yz89 — APPROVED** [review]

**Pull request merged**